### DISCUSSION TITLE
Focus throttling

### DISCUSSION THREAD

**Discussion by yusufyildirim** [react-native-tvos/react-native-tvos#667] [Feature Requests]
Currently there's no limit on how fast users can focus from one element to another. This is an issue for both UX and performance. If we can't control user's speed, we can't reliably provide good animations. If user spam arrow keys on their remote, the app has to catch up with their command and try to render countless items in lists (virtualization), in a short amount of time. This is a big burden on low-end devices. Also, we want the users to see all the content we provide, we don't want them to skip. So, we need a way to control focus change speed/interval.

This is how it looks when users spam their remote arrow keys on `react-native-tvos` applications by default:

https://github.com/react-native-tvos/react-native-tvos/assets/22980987/370495fa-8e5b-4f96-96e5-857353230225

Focus and scroll animation have hard time catching up with the speed of focused element. This is not desirable. Here's an example on how Youtube reacts when I spam right and left arrow keys:

https://github.com/react-native-tvos/react-native-tvos/assets/22980987/af9d26bb-cfef-465d-afcb-1856748588c3

There's certainly a limit they apply. Their approach feels a bit more different than what I'll propose but it should be fine.

I think we should introduce a prop called `focusThrottlingInterval` (please suggest a better name) for `TVFocusGuideView` to let developers configure focus throttling, specific to their application's needs.

```tsx
<TVFocusGuideView
  // ... other props
  focusThrottlingInterval={500} // in ms
/>
```

It should be configurable per `TVFocusGuideView` basis which should give a lot of flexibility to the developers. This way they can choose throttle focus on lists but not on the other parts of the app like menus, for example.

This is the most naive and straightforward way I could think of to provide a solution for this problem. I'll put a bit more thought into this but I need to get my hands dirty first to see what's viable or not.

**Comment by matias-sandell** [comment]
Great suggestion @yusufyildirim! I wonder if it would make sense to also add some kind of `focusThrottlingTrigger` (not a perfect name) which would be the number of focus changes before throttling kicks in?

This would be for those cases where I want to have my cake and eat it too, like a TV channel zapper UI, where the precision of focus changes are quite important in the beginning (very quick muscle memory clicks for three channels up or down) but not so relevant after ten or so clicks.

**Reply by yusufyildirim** [reply]
Yeah, I get that. I feel like we need something smarter here. Straight up throttling wouldn't feel intuitive I can imagine. That's why I ended with `I'll put a bit more thought into this but I need to get my hands dirty first to see what's viable or not.`. I'll see what we can do 😄 Just wanted to start a conversation around a simple idea

**Reply by douglowder** [reply]
Looked at this problem a bit, and I think what we want here is not so much throttling as "debouncing", where if too many events come in at once, we hold off processing them until there is a pause in the events.  This is commonly used to deal with user input where the user types too fast, for example... I think for focus debouncing, there are other optimizations we can try, for example, if too many focus events come in, we wait until they stop, but only allow the last few to fire since they are the only ones that will be relevant by then.

**Reply by yusufyildirim** [reply]
I'm positive that both throttling and debouncing will feel awkward if we implement them as is. Youtube does something based on scroll animation. They queue the events if animation is in progress and dequeue as animation ends. At least that's what I feel and it feels intuitive. We may need a bit more sophisticated approach here.

**Reply by douglowder** [reply]
Here's a first cut at doing this in the focus event handler... definitely improves the scrolling for long lists, but still needs some adjustments for sure.

```patch
diff --git a/node_modules/react-native/Libraries/Components/TV/TVFocusEventHandler.js b/node_modules/react-native/Libraries/Components/TV/TVFocusEventHandler.js
index 01f6e86..805b249 100644
--- a/node_modules/react-native/Libraries/Components/TV/TVFocusEventHandler.js
+++ b/node_modules/react-native/Libraries/Components/TV/TVFocusEventHandler.js
@@ -16,6 +16,33 @@ import {type EventSubscription} from '../../vendor/emitter/EventEmitter';
 import NativeTVNavigationEventEmitter from './NativeTVNavigationEventEmitter';
 import type {TVRemoteEvent} from '../../Types/CoreEventTypes';
 
+const debounceDelay = 300;
+const debounceMaxEvents = 5;
+
+// Debounces the focus events with a 300 ms debounce window
+// Only fire the last 5 events seen
+const debounce = (callback, data) => {
+  let timeoutId;
+  let queue = [];
+  return () => {
+    queue.push({
+      callback,
+      data,
+    });
+    if (queue.length > debounceMaxEvents) {
+      queue.shift();
+    }
+    clearTimeout(timeoutId);
+    timeoutId = setTimeout(() => {
+      while (queue.length > 0) {
+        const item = queue.shift();
+        item.callback(item.data);
+      }
+      queue = [];
+    }, debounceDelay);
+  };
+};
+
 class TVFocusEventHandler {
   __nativeTVNavigationEventListener: ?EventSubscription = null;
   __nativeTVNavigationEventEmitter: ?NativeEventEmitter<TVRemoteEvent> = null;
@@ -35,7 +62,8 @@ class TVFocusEventHandler {
       data => {
         const callback = this.__callbackMap.get(data.tag);
         if (callback) {
-          callback(data);
+          // callback(data);
+          debounce(callback, data)();
         }
       },
     );
```

**Comment by duveborg** [comment]
Any news on this? 👀
